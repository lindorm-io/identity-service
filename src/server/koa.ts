import { KoaApp } from "@lindorm-io/koa";
import { SERVER_PORT, TOKEN_ISSUER_MW_OPTIONS } from "../config";
import { appRoot, headless, identity } from "../route";
import { getMongoMiddleware, getWebKeyMiddleware, repositoryMiddleware } from "../middleware";
import { tokenIssuerMiddleware } from "@lindorm-io/koa-jwt";
import { winston } from "../logger";

export const koa = new KoaApp({
  logger: winston,
  port: SERVER_PORT,
});

koa.addMiddleware(getMongoMiddleware());
koa.addMiddleware(repositoryMiddleware);
koa.addMiddleware(getWebKeyMiddleware());
koa.addMiddleware(tokenIssuerMiddleware(TOKEN_ISSUER_MW_OPTIONS));

koa.addRoute("/", appRoot);
koa.addRoute("/headless", headless);
koa.addRoute("/identity", identity);
