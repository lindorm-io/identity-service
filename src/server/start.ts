import { koa } from "./koa";

(async () => {
  await koa.start();
})();
