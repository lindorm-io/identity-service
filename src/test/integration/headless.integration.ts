import MockDate from "mockdate";
import request from "supertest";
import { koa } from "../../server/koa";
import { inMemoryStore, setupIntegration, TEST_IDENTITY } from "../grey-box";
import { baseHash } from "@lindorm-io/core";

jest.mock("uuid", () => ({
  v4: jest.fn(() => "e3926ddb-ecaf-4f66-855a-d143af54953c"),
}));

MockDate.set("2020-01-01 08:00:00.000");

const basicAuth = baseHash("basic_auth_username:basic_auth_password");

describe("/headless", () => {
  beforeAll(async () => {
    await setupIntegration();
    koa.load();
  });

  test("POST /:id - should create", async () => {
    const response = await request(koa.callback())
      .post("/headless/0f9c57ad-e382-4a8b-b926-093c3a40f3f3")
      .set("Authorization", `Basic ${basicAuth}`)
      .set("X-Client-ID", "5c63ca22-6617-45eb-9005-7c897a25d375")
      .set("X-Correlation-ID", "5c63ca22-6617-45eb-9005-7c897a25d375")
      .expect(201);

    expect(response.body).toMatchSnapshot();
    expect(inMemoryStore).toMatchSnapshot();
  });

  test("POST /:id - should find", async () => {
    const response = await request(koa.callback())
      .post(`/headless/${TEST_IDENTITY.id}`)
      .set("Authorization", `Basic ${basicAuth}`)
      .set("X-Client-ID", "5c63ca22-6617-45eb-9005-7c897a25d375")
      .set("X-Correlation-ID", "5c63ca22-6617-45eb-9005-7c897a25d375")
      .expect(201);

    expect(response.body).toMatchSnapshot();
    expect(inMemoryStore).toMatchSnapshot();
  });
});
