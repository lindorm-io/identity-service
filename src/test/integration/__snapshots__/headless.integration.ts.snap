// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/headless POST /:id - should create 1`] = `
Object {
  "created": "2020-01-01T07:00:00.000Z",
  "updated": "2020-01-01T07:00:00.000Z",
}
`;

exports[`/headless POST /:id - should create 2`] = `
Object {
  "identity": Object {
    "display_name": Object {
      "data": Array [
        Object {
          "created": 2020-01-01T07:00:00.000Z,
          "events": Array [
            Object {
              "date": 2020-01-01T07:00:00.000Z,
              "name": "display_name_created",
              "payload": Object {
                "name": "displayName",
                "numbers": Array [
                  1234,
                ],
              },
            },
          ],
          "id": "e3926ddb-ecaf-4f66-855a-d143af54953c",
          "name": "displayName",
          "numbers": Array [
            1234,
          ],
          "updated": 2020-01-01T07:00:00.000Z,
          "version": 0,
        },
      ],
      "indices": Array [
        Object {
          "index": Object {
            "id": 1,
          },
          "options": Object {
            "unique": true,
          },
        },
      ],
    },
    "identity": Object {
      "data": Array [
        Object {
          "address": Object {
            "country": "country",
            "locality": "locality",
            "postalCode": "postalCode",
            "region": "region",
            "streetAddress": "streetAddress",
          },
          "birthDate": "2000-01-01",
          "created": 2020-01-01T07:00:00.000Z,
          "displayName": Object {
            "name": "displayName",
            "number": 1234,
          },
          "events": Array [
            Object {
              "date": 2020-01-01T07:00:00.000Z,
              "name": "identity_created",
              "payload": Object {
                "address": Object {
                  "country": "country",
                  "locality": "locality",
                  "postalCode": "postalCode",
                  "region": "region",
                  "streetAddress": "streetAddress",
                },
                "birthDate": "2000-01-01",
                "displayName": Object {
                  "name": "displayName",
                  "number": 1234,
                },
                "familyName": "familyName",
                "gender": "gender",
                "givenName": "givenName",
                "gravatar": "https://gravatar.url/",
                "locale": "sv-SE",
                "middleName": "middleName",
                "nickname": "nickname",
                "phoneNumber": "+46700000000",
                "phoneNumberVerified": true,
                "picture": "https://picture.url/",
                "preferredUsername": "preferredUsername",
                "profile": "https://profile.url/",
                "username": "username",
                "website": "https://website.url/",
                "zoneInfo": "Europe/Stockholm",
              },
            },
          ],
          "familyName": "familyName",
          "gender": "gender",
          "givenName": "givenName",
          "gravatar": "https://gravatar.url/",
          "id": "e3926ddb-ecaf-4f66-855a-d143af54953c",
          "locale": "sv-SE",
          "middleName": "middleName",
          "nickname": "nickname",
          "phoneNumber": "+46700000000",
          "phoneNumberVerified": true,
          "picture": "https://picture.url/",
          "preferredUsername": "preferredUsername",
          "profile": "https://profile.url/",
          "updated": 2020-01-01T07:00:00.000Z,
          "username": "username",
          "version": 0,
          "website": "https://website.url/",
          "zoneInfo": "Europe/Stockholm",
        },
        Object {
          "address": null,
          "birthDate": null,
          "created": 2020-01-01T07:00:00.000Z,
          "displayName": null,
          "events": Array [
            Object {
              "date": 2020-01-01T07:00:00.000Z,
              "name": "identity_created",
              "payload": Object {
                "address": null,
                "birthDate": null,
                "displayName": null,
                "familyName": null,
                "gender": null,
                "givenName": null,
                "gravatar": null,
                "locale": null,
                "middleName": null,
                "nickname": null,
                "phoneNumber": null,
                "phoneNumberVerified": false,
                "picture": null,
                "preferredUsername": null,
                "profile": null,
                "username": null,
                "website": null,
                "zoneInfo": null,
              },
            },
          ],
          "familyName": null,
          "gender": null,
          "givenName": null,
          "gravatar": null,
          "id": "0f9c57ad-e382-4a8b-b926-093c3a40f3f3",
          "locale": null,
          "middleName": null,
          "nickname": null,
          "phoneNumber": null,
          "phoneNumberVerified": false,
          "picture": null,
          "preferredUsername": null,
          "profile": null,
          "updated": 2020-01-01T07:00:00.000Z,
          "username": null,
          "version": 0,
          "website": null,
          "zoneInfo": null,
        },
      ],
      "indices": Array [
        Object {
          "index": Object {
            "id": 1,
          },
          "options": Object {
            "unique": true,
          },
        },
      ],
    },
  },
}
`;

exports[`/headless POST /:id - should find 1`] = `
Object {
  "created": "2020-01-01T07:00:00.000Z",
  "updated": "2020-01-01T07:00:00.000Z",
}
`;

exports[`/headless POST /:id - should find 2`] = `
Object {
  "identity": Object {
    "display_name": Object {
      "data": Array [
        Object {
          "created": 2020-01-01T07:00:00.000Z,
          "events": Array [
            Object {
              "date": 2020-01-01T07:00:00.000Z,
              "name": "display_name_created",
              "payload": Object {
                "name": "displayName",
                "numbers": Array [
                  1234,
                ],
              },
            },
          ],
          "id": "e3926ddb-ecaf-4f66-855a-d143af54953c",
          "name": "displayName",
          "numbers": Array [
            1234,
          ],
          "updated": 2020-01-01T07:00:00.000Z,
          "version": 0,
        },
      ],
      "indices": Array [
        Object {
          "index": Object {
            "id": 1,
          },
          "options": Object {
            "unique": true,
          },
        },
      ],
    },
    "identity": Object {
      "data": Array [
        Object {
          "address": Object {
            "country": "country",
            "locality": "locality",
            "postalCode": "postalCode",
            "region": "region",
            "streetAddress": "streetAddress",
          },
          "birthDate": "2000-01-01",
          "created": 2020-01-01T07:00:00.000Z,
          "displayName": Object {
            "name": "displayName",
            "number": 1234,
          },
          "events": Array [
            Object {
              "date": 2020-01-01T07:00:00.000Z,
              "name": "identity_created",
              "payload": Object {
                "address": Object {
                  "country": "country",
                  "locality": "locality",
                  "postalCode": "postalCode",
                  "region": "region",
                  "streetAddress": "streetAddress",
                },
                "birthDate": "2000-01-01",
                "displayName": Object {
                  "name": "displayName",
                  "number": 1234,
                },
                "familyName": "familyName",
                "gender": "gender",
                "givenName": "givenName",
                "gravatar": "https://gravatar.url/",
                "locale": "sv-SE",
                "middleName": "middleName",
                "nickname": "nickname",
                "phoneNumber": "+46700000000",
                "phoneNumberVerified": true,
                "picture": "https://picture.url/",
                "preferredUsername": "preferredUsername",
                "profile": "https://profile.url/",
                "username": "username",
                "website": "https://website.url/",
                "zoneInfo": "Europe/Stockholm",
              },
            },
          ],
          "familyName": "familyName",
          "gender": "gender",
          "givenName": "givenName",
          "gravatar": "https://gravatar.url/",
          "id": "e3926ddb-ecaf-4f66-855a-d143af54953c",
          "locale": "sv-SE",
          "middleName": "middleName",
          "nickname": "nickname",
          "phoneNumber": "+46700000000",
          "phoneNumberVerified": true,
          "picture": "https://picture.url/",
          "preferredUsername": "preferredUsername",
          "profile": "https://profile.url/",
          "updated": 2020-01-01T07:00:00.000Z,
          "username": "username",
          "version": 0,
          "website": "https://website.url/",
          "zoneInfo": "Europe/Stockholm",
        },
        Object {
          "address": null,
          "birthDate": null,
          "created": 2020-01-01T07:00:00.000Z,
          "displayName": null,
          "events": Array [
            Object {
              "date": 2020-01-01T07:00:00.000Z,
              "name": "identity_created",
              "payload": Object {
                "address": null,
                "birthDate": null,
                "displayName": null,
                "familyName": null,
                "gender": null,
                "givenName": null,
                "gravatar": null,
                "locale": null,
                "middleName": null,
                "nickname": null,
                "phoneNumber": null,
                "phoneNumberVerified": false,
                "picture": null,
                "preferredUsername": null,
                "profile": null,
                "username": null,
                "website": null,
                "zoneInfo": null,
              },
            },
          ],
          "familyName": null,
          "gender": null,
          "givenName": null,
          "gravatar": null,
          "id": "0f9c57ad-e382-4a8b-b926-093c3a40f3f3",
          "locale": null,
          "middleName": null,
          "nickname": null,
          "phoneNumber": null,
          "phoneNumberVerified": false,
          "picture": null,
          "preferredUsername": null,
          "profile": null,
          "updated": 2020-01-01T07:00:00.000Z,
          "username": null,
          "version": 0,
          "website": null,
          "zoneInfo": null,
        },
      ],
      "indices": Array [
        Object {
          "index": Object {
            "id": 1,
          },
          "options": Object {
            "unique": true,
          },
        },
      ],
    },
  },
}
`;
